components/Calendar.js

```diff
@@ -160,63 +192,88 @@ class Calendar extends Component {
+        const existingStartDate = this.props.multiRangeStartDates[i];
+        const existingEndDate = moment(existingStartDate).add(this.props.multiRangeLength - 1, 'days');
+        if (endDate.isSameOrAfter(existingStartDate, 'days') &&
+            endDate.isSameOrBefore(existingEndDate, 'days')) return;
+      }
+
+      /* Checkpoint 3: Adding new selection and run callback */
+      this.props.onDateSelect && this.props.onDateSelect(date);
+      this.props.onMultiRangeStartDateSelect && this.props.onMultiRangeStartDateSelect(date);
+      if (this.props.selectedDate === undefined) {
+        this.props.multiRangeStartDates.push(date);
+        this.setState({ selectedMoment: date, selectedMultiRangeStartDate: date });
+      }
+      return;
+    }
+    /* Regular Mode */
     if (this.props.selectedDate === undefined) {
       this.setState({ selectedMoment: date });
     }
     let rangeStart = this.state.rangeStart;
     let rangeEnd = this.state.rangeEnd;
-    let selectStartDate = this.state.selectStartDate;
+    const selectStartDate = this.state.selectStartDate;
     let firstSelection = this.state.firstSelection;
-      if (this.props.rangeEnabled) {
-        if (firstSelection && selectStartDate) {
-          rangeEnd = date;
-          rangeStart = date;
-        } else if (firstSelection && !selectStartDate && date.isAfter(rangeStart)) {
-          rangeEnd = date;
-          firstSelection = false;
-        } else if (selectStartDate && date.isBefore(rangeEnd)) {
-          rangeStart = date;
-        } else if (!selectStartDate && date.isAfter(rangeStart)) {
-          rangeEnd = date;
-        } else {
-          return;
-        }
-        this.setState({
-          rangeStart,
-          rangeEnd,
-          firstSelection,
-        });
-        // If selected day is rangeEnd, run onEndDateSelect callback
-        this.state.rangeEnd != rangeEnd ?
-          this.props.onEndDateSelect && this.props.onEndDateSelect(date ? date.format() : null,
-            null,
-            rangeEnd ? rangeEnd.format() : null ) :
-          null;
-        // If selected day is rangeStart, run onStartDateSelect callback
-        this.state.rangeStart != rangeStart ?
-          this.props.onStartDateSelect && this.props.onStartDateSelect(date ? date.format() : null,
-            rangeStart ? rangeStart.format() : null
-          ) :
-          null;
-
+    if (this.props.rangeEnabled) {
+      if (firstSelection && selectStartDate) {
+        rangeEnd = date;
+        rangeStart = date;
+      } else if (firstSelection && !selectStartDate && date.isAfter(rangeStart)) {
+        rangeEnd = date;
+        firstSelection = false;
+      } else if (selectStartDate && date.isBefore(rangeEnd)) {
+        rangeStart = date;
+      } else if (!selectStartDate && date.isAfter(rangeStart)) {
+        rangeEnd = date;
+      } else {
         return;
       }
+      this.setState({
```

Comment:
Since setState has a callback param option, would having all the stuff after setState be in that callback resolve the confusion?